### DISCUSSION TITLE
How to create a function 'add_constraints' ?

### DISCUSSION THREAD

**Discussion by longevite8** [skfolio/skfolio#65] [Q&A] [answered]
Hi everyone,

I need to create a function add_contraints for optimizing portfolio with given expected return, please show me a example !

Thank you so much!

**Comment by HugoDelatte** [comment, marked as answer]
Hello,

I agree that's missing from the doc. I'll add a proper example.

Here is how you can add custom constraints (but note that you can already achieve a lot by using `min_weights`, `max_weights`, 
`budget`, `min_budget`, `max_budget`, `max_short`, `max_long`, `linear_constraints`, `groups`, `left_inequality`, `right_inequality`):


```python
from skfolio.datasets import load_sp500_dataset
from skfolio.optimization import MeanRisk
from skfolio.preprocessing import prices_to_returns

prices = load_sp500_dataset()
X = prices_to_returns(prices)


def custom_constraints(w):
    return [w <= 0.1, w >= 0.01]


model = MeanRisk(add_constraints=custom_constraints)
model.fit(X)
print(model.weights_)
```

Let me know if that solves the issue, otherwise could you tell me the constraint you need to implement?

**Reply by longevite8** [reply]
Thanks for your rapid response!

In the case, I need a constraint that expected return of portfolio must be a given value. Can I do that with following code ?

prices = load_sp500_dataset()
X = prices_to_returns(prices)

R : matrix of return of assets, and exp_value is expected return of portfolio
def  expected_return_constraints(w, R, exp_value):
      return [ w*R == exp_value]

exp_return = 0.15 
model = MeanRisk(add_constraints=expected_return_constraints(X, exp_return))

Thanks for your support!

**Reply by HugoDelatte** [reply]
In that case it's even simpler by using `min_return`:

For example, to target an expected return of 20% p.a.:

```python
model = MeanRisk(min_return=0.2 / 252)
model.fit(X)
print(model.weights_)

assert abs(model.predict(X).mean - 0.2 / 252) < 1e-10
```

**Reply by longevite8** [reply]
Ok, if possible, can you add max_return for the problem with expected return in a value range [min_return, max_return], similarly with  min_standard_deviation or min_variance ... for optimal portfolio with expected standard deviation or variance in a value range ?.

Thank you so much!

**Comment by HugoDelatte** [comment]
In most cases that would be an anti-pattern for portfolio optimization. If you set a upper constrain on the expected returns and minimize the variance, in best case you will end-up with the minimum variance portfolio (so the constraint is useless),  and in worst case you may even end-up with non-pareto optimal portfolio.
If an user has a very good reason to do so, it can be achieved using the `add_contraints` parameter as you did but we shouldn't facilitate it I think.

**Reply by longevite8** [reply]
Ok, thanks Hugo

**Reply by kevingivens** [reply]
Hello and thanks for the wonderful library.  Is there a way to add a class method as a custom constraint? Do I need to inherit from `ConvexOptimization`?  Thanks

**Reply by HugoDelatte** [reply]
You can indirectly by using the argument `add_constraints` which accept a function taking for argument the weight parameter and returns a list of cvxpy constraints (as shown above).
If you provide a second argument, it will be associated with the class instance so you can use it like a method (I've just realised that this explanation is missing in the docstring):

```python
from skfolio.datasets import load_sp500_dataset
from skfolio.optimization import MeanRisk
from skfolio.preprocessing import prices_to_returns

prices = load_sp500_dataset()
X = prices_to_returns(prices)


def custom_constraints(w, obj):
    return [w >= obj.budget / 20]


model = MeanRisk(add_constraints=custom_constraints)
model.fit(X)
print(model.weights_)
```



Could you kindly provide more details on the custom constrain you want to add?
Because if it something standard we could add it natively.

**Reply by kevingivens** [reply]
Thanks a lot for your help.  I was thinking of non-linear constraints like `cp.sum(cp.square(w)) < limit`.
